### PR TITLE
Patch for UHD HDR10 and HDR10+

### PR DESCRIPTION
**Pull request by jcdr428**
I've tested on few SDR and HDR10 Blu-rays without issues.

The patch would need to be thoroughly tested by some other users (eg offer an alternative build for testing) before integration to Master.

### CODE REVIEW

**Comment by justdan96** [comment]
It looks like these changes conflict with some of the changes @xavery introduced in #67 

Can you please take a look at how to resolve these conflicts?

**Comment by jcdr428** [comment]
Grr... I have to replace most "virtual" with "overrides". Will do !!

**Comment by jcdr428** [comment]
@justdan96 the only conflicts left are the integration of functions getStreamHDR() in the four StreamReaders. Nothing else I can do on my side.

File: tsMuxer/BufferedReader.h [hunk 1]
```diff
@@ -5,6 +5,7 @@
 #include <map>
 #include <types/types.h>
 #include <system/terminatablethread.h>
+#include <mutex>
```
Comment (xavery) [inline tsMuxer/BufferedReader.h]: `safequeue.h` that's included in the following line already includes `<mutex>` - are you sure this is needed here?

File: tsMuxer/blurayHelper.cpp [hunk 2]
```diff
@@ -165,22 +166,29 @@ bool BlurayHelper::writeBluRayFiles(bool usedBlackPL, int mplsNum, int blankNum,
...
     }
     else if (m_dt == UHD_BLURAY) {
         bdMovieObjectData[5] = bdIndexData[5] = '3';
-        fileSize = 0x78;
+        fileSize = 0x9C; // add 36 bytes for HDR data extension
+        bdIndexData[15] = 0x78; // start address of HDR data extension
+        emptyCommand = bdIndexData + 0x78;
+        memcpy(emptyCommand, "\x00\x00\x00\x20\x00\x00\x00\x18\x00\x00\x00\x01"
+                             "\x00\x03\x00\x01\x00\x00\x00\x18\x00\x00\x00\x0C"
+                             "\x00\x00\x00\x08\x51\x00\x00\x00\x00\x00\x00\x00", 36); // HDR data extension
```
Comment (xavery) [inline tsMuxer/blurayHelper.cpp]: If you replace this string literal with an array of `uint8_t`, you'll get less code and will be able to use `sizeof` instead of hard-coding the size of the array.

File: tsMuxer/blurayHelper.cpp [hunk 3]
```diff
@@ -165,22 +166,29 @@ bool BlurayHelper::writeBluRayFiles(bool usedBlackPL, int mplsNum, int blankNum,
...
+        bdIndexData[0x96] = (byte)HDR10_metadata[0]; // HDR flags
```
Comment (xavery) [inline tsMuxer/blurayHelper.cpp]: `std::byte` is a C++17 feature, and this code rightly causes a compilation failure with Clang. `static_cast` to `uint8_t` instead.

File: tsMuxer/abstractStreamReader.h [hunk 4]
```diff
@@ -32,6 +32,7 @@ class AbstractStreamReader: public BaseAbstractStreamReader {
 	virtual int readPacket(AVPacket& avPacket) = 0;
 	virtual int flushPacket(AVPacket& avPacket) = 0;
 	virtual int getTSDescriptor(uint8_t* dstBuff) {return 0;}
+	virtual int getStreamHDR() const { return 0; }
```
Comment (xavery) [inline tsMuxer/abstractStreamReader.h]: This causes `getStreamHDR` not to be abstract in all the classes deriving from `AbstractStreamReader`, notably `MPEGStreamReader` which redeclares the function as abstract. Is this intentional?

File: tsMuxer/h264StreamReader.h [hunk 5]
```diff
@@ -3,6 +3,7 @@
 
 #include <map>
 #include <iostream>
+#include <algorithm>
```
Comment (xavery) [inline tsMuxer/h264StreamReader.h]: There is no code in the header which directly uses stuff from `<algorithm>` - why was this moved here from the source file?

File: tsMuxer/hevc.cpp [hunk 6]
```diff
@@ -807,6 +809,74 @@ int HevcPpsUnit::deserialize()
...
+		return rez;
+	try
+	{
+		int payloadType;
+		int payloadSize;
+		int nbyte;
+		do {
+			payloadType = 0;
+			payloadSize = 0;
+			nbyte = 0xFF;
```
Comment (xavery) [inline tsMuxer/hevc.cpp]: Any reason why these variables are initialised later (unconditionally, since the `do` block is going to be entered either way) instead of at declaration time? If not, just initialise them while declaring.

File: tsMuxer/hevc.h [hunk 7]
```diff
@@ -3,6 +3,8 @@
 
 #include "nalUnits.h"
 
+extern int HDR10_metadata[7];
```
Comment (xavery) [inline tsMuxer/hevc.h]: This is a global variable that's set by parts of `hevc.cpp` and then read by `blurayHelper.cpp`. Would it be possible to reduce its scope, for example to a member variable of the `BlurayHelper` class?

File: tsMuxer/hevcStreamReader.h [hunk 8]
```diff
@@ -41,6 +42,7 @@ class HEVCStreamReader: public MPEGStreamReader
     HevcVpsUnit* m_vps;
     HevcSpsUnit* m_sps;
     HevcPpsUnit* m_pps;
+	HevcSeiUnit* m_sei;
```
Comment (xavery) [inline tsMuxer/hevcStreamReader.h]: Any reason why it's not a `std::unique_ptr`?

File: tsMuxer/metaDemuxer.h [hunk 9]
```diff
@@ -10,6 +10,8 @@
 #include <set>
 #include <chrono>
 #include <queue>
+#include <iomanip>
+#include <chrono>
```
Comment (xavery) [inline tsMuxer/metaDemuxer.h]: Not only is `<chrono>` already included by this header, but there is no code that uses `<iomanip>` which would require its inclusion here.

File: tsMuxer/blurayHelper.cpp (hunk 3 above)
Comment (xavery) [reply tsMuxer/blurayHelper.cpp]: Looks like [GCC doesn't like this either](https://github.com/jcdr428/tsMuxer/commit/d8d9669973af16b4e6c64eb6b0d08686327ad099/checks#step:5:62).

File: tsMuxer/BufferedReader.h (hunk 1 above)
Comment (jcdr428) [reply tsMuxer/BufferedReader.h]: @xavery you are correct. I didn't apply the changes to libmediation, so safequeue.h was still referencing to system/mutex.h

File: tsMuxer/abstractStreamReader.h (hunk 4 above)
Comment (jcdr428) [reply tsMuxer/abstractStreamReader.h]: @xavery you're correct: removed from AbstractStreamReader. First experience with C++, I'm still unfamiliar with inheritance...

File: tsMuxer/hevc.cpp (hunk 6 above)
Comment (jcdr428) [reply tsMuxer/hevc.cpp]: @xavery once again, so true... Thanks for correcting.

File: tsMuxer/metaDemuxer.h (hunk 9 above)
Comment (jcdr428) [reply tsMuxer/metaDemuxer.h]: Both removed !

File: tsMuxer/h264StreamReader.h (hunk 5 above)
Comment (jcdr428) [reply tsMuxer/h264StreamReader.h]: @xavery This is correct. Reference removed.

File: tsMuxer/hevc.h (hunk 7 above)
Comment (jcdr428) [reply tsMuxer/hevc.h]: @xavery I've tried for hours to set the variable in a specific class. Then it has to be passed from class to class, I didn't manage to do this and ended up setting the global variable... Maybe someone else will manage to do this.

File: tsMuxer/hevcStreamReader.h (hunk 8 above)
Comment (jcdr428) [reply tsMuxer/hevcStreamReader.h]: @xavery not sure what you mean here. I used the same struct for sei as for vps, sps and pps.

File: tsMuxer/hevc.h (hunk 7 above)
Comment (xavery) [reply tsMuxer/hevc.h]: I understand - sadly, I also have no idea on how to get this through in a more civilised way, so I'll just let it go.

File: tsMuxer/hevcStreamReader.h (hunk 8 above)
Comment (xavery) [reply tsMuxer/hevcStreamReader.h]: There is no reason for this to be a raw pointer : since C++11, we can use the `std::unique_ptr` class, which guarantees that the pointer is never "accidentally" copied and its ownership is always tied to the `unique_ptr` object. Existing code doesn't do this, because it was written before these classes were available.

**Comment by xavery** [comment]
@jcdr428 I'm fine with merging these changes once you're confident that enough testing has been done.

**Comment by jcdr428** [comment]
@xavery thanks. jdobbs and I have tested this HDR10 patch quite a lot, and it is working well.
tymoxa has also tested and has reported incorrectly formed TS packets / descriptors which are not linked to this patch -I have another patch ready to be pushed for correcting the HEVC and AVC descriptors.

So in brief, yes I am confident that enough testing has been done to merge this change.
